### PR TITLE
aya: add archs powerpc64 and s390x to aya

### PR DESCRIPTION
**Pull request by Billy99**
bpfman, a project using aya, has a requirement to support powerpc64 and s390x architectures. Adding these two architectures to aya.

### CODE REVIEW

File: ebpf/aya-ebpf/src/args.rs [hunk 1]
```diff
@@ -168,6 +173,36 @@ impl<T> FromPtRegs for *const T {
...
+        if n <= 7 {
+            unsafe { bpf_probe_read(&ctx.gpr[3 + n]).map(|v| v as *const _).ok() }
+        } else {
+            None
+        }
+    }
+
+    fn from_retval(ctx: &pt_regs) -> Option<Self> {
+        unsafe { bpf_probe_read(&ctx.gpr[3]).map(|v| v as *const _).ok() }
+    }
```
Comment (Billy99) [inline ebpf/aya-ebpf/src/args.rs:188]: Just to verify that I did this section correctly. For `from_retval()` for PowerPC, I used https://github.com/libbpf/libbpf/blob/master/src/bpf_tracing.h#L170 as reference and returning the value of `__PT_RC_REG`.

I wasn't sure about this. X86 and ARM64 are returning `__PT_RC_REG`, but RISCV appears to be returning `__PT_RET_REG`. Is RISCV coded wrong or is there something I'm missing?
X86: https://github.com/libbpf/libbpf/blob/master/src/bpf_tracing.h#L150
ARM64: https://github.com/libbpf/libbpf/blob/master/src/bpf_tracing.h#L250
RISCV: https://github.com/libbpf/libbpf/blob/master/src/bpf_tracing.h#L374

File: ebpf/aya-ebpf/src/args.rs [hunk 2]
```diff
@@ -168,6 +173,36 @@ impl<T> FromPtRegs for *const T {
...
+        if n <= 4 {
+            unsafe { bpf_probe_read(&ctx.gprs[2 + n]).map(|v| v as *const _).ok() }
+        } else {
+            None
+        }
+    }
+
+    fn from_retval(ctx: &pt_regs) -> Option<Self> {
+        unsafe { bpf_probe_read(&ctx.gprs[2]).map(|v| v as *const _).ok() }
+    }
```
Comment (Billy99) [inline ebpf/aya-ebpf/src/args.rs:203]: Just to verify that I did this section correctly. For `from_retval()` for s390x, I used https://github.com/libbpf/libbpf/blob/master/src/bpf_tracing.h#L188 as reference and returning the value of `__PT_RC_REG`. See PowerPC comment regarding `__PT_RC_REG` vs `__PT_RET_REG`.

**Comment by dave-tucker** [comment]
@Billy99 I've opened #975 which just adds the bits to xtask.
Once that merges I'll trigger the codegen job which will generate the necessary bindings in CI.
You can then rebase on that and see if that resolves the issue 🤞 

Otherwise ISTR there were issues with s390x emulation in bimft_misc.
Let me see if I can find a citation for that... and hopefully we can try and work around it 🙏

**Comment by dave-tucker** [comment]
@Billy99 if you don't mind rebasing that would be awesome!
Codegen has been done in CI to avoid any of the weird issues we've hit in the past with codegen - which may or may not be the cause of your CI issue. At least this way we can rule that out.

**Review by astoycos — APPROVED** [review]
This looks good to me, I hate that we have to manually maintain BE and LE versions of each of those test byte streams but I'm not sure how to automagically flip them unless we explicitly match them to the types they're getting decoded to 

CC @tamird Can you please give this a look as well?

**Comment by tamird** [comment]
@astoycos couldn't we keep only the LE and conditionally compile a shadowing variable declaration that reverses the endianness at runtime?

```rust
let foo = ...;
#[cfg(...)]
let foo = le_to_be(foo);
```

**Comment by dave-tucker** [comment]
> @astoycos couldn't we keep only the LE and conditionally compile a shadowing variable declaration that reverses the endianness at runtime?
> 
> ```rust
> let foo = ...;
> #[cfg(...)]
> let foo = le_to_be(foo);
> ```

Unfortunately not because swapping endianness requires knowing the inner workings of the struct represented by the byte array.

```rust
use std::io::Cursor;
use byteorder::{BigEndian, LittleEndian, ReadBytesExt};

fn main() {

    // byte array represents 2 x u16s
    let mut le_bytes = Cursor::new(vec![0, 3 ,5 ,2]);
    
    // naively swap byte order
    let mut naive_be_bytes = Cursor::new(vec![2, 5, 3, 0]);
    
    // swap byte order, knowing this is 2 x u16s
    let mut be_bytes = Cursor::new(vec![3, 0, 2, 5]);
    
    assert_eq!(768, le_bytes.read_u16::<LittleEndian>().unwrap());
    assert_eq!(768, be_bytes.read_u16::<BigEndian>().unwrap());
    assert_eq!(768, naive_be_bytes.read_u16::<BigEndian>().unwrap()); // this assertion will fail
}
```

[Rust playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f78e4eb7f146cae6799d165a3fe5a269)

**Comment by tamird** [comment]
You just need to know the element size, right?

**Comment by Billy99** [comment]
> You just need to know the element size, right?

Yes, but most are structs like btf_header. So its:
u16, u8, u8, u32, u32, u32 ... u8, u8, ...
Where the u8 start after header.str_off.

**Comment by tamird** [comment]
I see. No free lunch then - you could implement a trait and use destructuring to get some compiler help but otherwise writing decoded literals rather than encoded seems a reasonable option.

**Comment by Billy99** [comment]
Thanks, it's not ideal, but what we have for now.

My bigger concern was with the second commit and if I did the PowerPC and s390x bits correctly, specifically in `ebpf/aya-ebpf/src/args.rs`.

File: ebpf/aya-ebpf/src/args.rs [hunk 3]
```diff
@@ -1,8 +1,13 @@
...
+    bpf_target_arch = "powerpc64"
+))]
 use crate::bindings::pt_regs;
-#[cfg(bpf_target_arch = "aarch64")]
+// aarch64 uses user_pt_regs instead of pt_regs
+#[cfg(any(bpf_target_arch = "aarch64", bpf_target_arch = "s390x"))]
 use crate::bindings::user_pt_regs as pt_regs;
+// riscv64 uses user_regs_struct instead of pt_regs
 #[cfg(bpf_target_arch = "riscv64")]
 use crate::bindings::user_regs_struct as pt_regs;
```
[the remark this answers is not in the document]
Comment (dave-tucker) [reply ebpf/aya-ebpf/src/args.rs:12]: LGTM

File: ebpf/aya-ebpf/src/args.rs [hunk 4]
```diff
@@ -168,6 +173,36 @@ impl<T> FromPtRegs for *const T {
...
 
+#[cfg(bpf_target_arch = "powerpc64")]
+impl<T> FromPtRegs for *const T {
+    fn from_argument(ctx: &pt_regs, n: usize) -> Option<Self> {
+        if n <= 7 {
+            unsafe { bpf_probe_read(&ctx.gpr[3 + n]).map(|v| v as *const _).ok() }
+        } else {
+            None
+        }
+    }
```
[the remark this answers is not in the document]
Comment (dave-tucker) [reply ebpf/aya-ebpf/src/args.rs:184]: Also LGTM

File: ebpf/aya-ebpf/src/args.rs (hunk 1 above)
Comment (dave-tucker) [reply ebpf/aya-ebpf/src/args.rs:188]: Also LGTM

File: ebpf/aya-ebpf/src/args.rs [hunk 5]
```diff
@@ -168,6 +173,36 @@ impl<T> FromPtRegs for *const T {
...
+
+#[cfg(bpf_target_arch = "s390x")]
+impl<T> FromPtRegs for *const T {
+    fn from_argument(ctx: &pt_regs, n: usize) -> Option<Self> {
+        if n <= 4 {
+            unsafe { bpf_probe_read(&ctx.gprs[2 + n]).map(|v| v as *const _).ok() }
+        } else {
+            None
+        }
+    }
```
[the remark this answers is not in the document]
Comment (dave-tucker) [reply ebpf/aya-ebpf/src/args.rs:199]: Also LGTM

File: ebpf/aya-ebpf/src/args.rs (hunk 2 above)
Comment (dave-tucker) [reply ebpf/aya-ebpf/src/args.rs:203]: Also LGTM

**Review by dave-tucker — APPROVED** [review]

File: aya-obj/src/obj.rs [hunk 6]
```diff
@@ -1594,22 +1595,33 @@ mod tests {
...
-            ".BTF",
-            &[
-                159, 235, 1, 0, 24, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
-            ],
-            None,
-        ))
-        .unwrap();
+
+        let btf = Btf::new();
+        //_ = btf.add_string("");
```
Comment (vadorovsky) [inline aya-obj/src/obj.rs:1600]: Can we remove this line?

Comment (Billy99) [reply aya-obj/src/obj.rs]: Done

**Review by alessandrod — APPROVED** [review]
great work!

**Pull request merged**